File: src/ray.ts
import * as V from './vec3.ts'

export type Ray = { origin: V.Vec3; direction: V.Vec3 }

export const ray =
  (origin: V.Vec3) => (direction: V.Vec3) =>
  ({ origin, direction })

export const at =
  (t: number) => ({ origin, direction }: Ray) =>
  V.add(origin)(V.multiply(t)(direction))

File: src/ray.rs
use crate::Vec3;

pub struct Ray<'a, 'b> {
    pub origin: &'a Vec3,
    pub direction: &'b Vec3,
}

impl Ray<'_, '_> {
    fn at(&self, t: f64) -> Vec3 {
        self.origin + &(self.direction * t)
    }
}
